File: backend/app/core/permissions.py
from app.models.user import UserRole

class Permissions:
    @staticmethod
    def can_create_issue(user_role: UserRole) -> bool:
        return user_role in [UserRole.ADMIN, UserRole.MAINTAINER, UserRole.REPORTER]
    
    @staticmethod
    def can_view_all_issues(user_role: UserRole) -> bool:
        return user_role in [UserRole.ADMIN, UserRole.MAINTAINER]
    
    @staticmethod
    def can_edit_any_issue(user_role: UserRole) -> bool:
        return user_role in [UserRole.ADMIN, UserRole.MAINTAINER]
    
    @staticmethod
    def can_delete_any_issue(user_role: UserRole) -> bool:
        return user_role == UserRole.ADMIN
    
    @staticmethod
    def can_assign_issues(user_role: UserRole) -> bool:
        return user_role in [UserRole.ADMIN, UserRole.MAINTAINER]
    
    @staticmethod
    def can_manage_users(user_role: UserRole) -> bool:
        return user_role == UserRole.ADMIN
    
    @staticmethod
    def can_view_issue(user_role: UserRole, issue_reporter_id: str, user_id: str) -> bool:
        if user_role in [UserRole.ADMIN, UserRole.MAINTAINER]:
            return True
        return issue_reporter_id == user_id
    
    @staticmethod
    def can_edit_issue(user_role: UserRole, issue_reporter_id: str, user_id: str) -> bool:
        if user_role in [UserRole.ADMIN, UserRole.MAINTAINER]:
            return True
        return issue_reporter_id == user_id

File: frontend/src/utils/permissions.ts
import { UserRole } from '../types';
import { rolePermissions } from './mockData';

export const hasPermission = (userRole: UserRole, permission: keyof typeof rolePermissions.ADMIN): boolean => {
  return rolePermissions[userRole][permission] || false;
};

export const canEditIssue = (userRole: UserRole, issueReporterId: string, userId: string): boolean => {
  if (hasPermission(userRole, 'canEditAnyIssue')) {
    return true;
  }
  return issueReporterId === userId;
};

export const canDeleteIssue = (userRole: UserRole, issueReporterId: string, userId: string): boolean => {
  if (hasPermission(userRole, 'canDeleteAnyIssue')) {
    return true;
  }
  return userRole === 'REPORTER' && issueReporterId === userId;
};

export const canViewIssue = (userRole: UserRole, issueReporterId: string, userId: string): boolean => {
  if (hasPermission(userRole, 'canViewAllIssues')) {
    return true;
  }
  return issueReporterId === userId;
};